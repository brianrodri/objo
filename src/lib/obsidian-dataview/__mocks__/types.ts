import { vi } from "vitest";

export const getAPI = vi.fn();
export const isPluginEnabled = vi.fn();
