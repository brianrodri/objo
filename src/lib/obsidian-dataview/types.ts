import { DateTime } from "luxon";
import type { STask } from "obsidian-dataview/lib/data-model/serialized/markdown";

export { DataviewApi, getAPI, isPluginEnabled } from "obsidian-dataview";
export type { SMarkdownPage as DataviewMarkdownPage } from "obsidian-dataview/lib/data-model/serialized/markdown";

export interface DataviewMarkdownTask extends STask {
    created?: DateTime;
    due?: DateTime;
    completion?: DateTime;
    start?: DateTime;
    scheduled?: DateTime;
}
