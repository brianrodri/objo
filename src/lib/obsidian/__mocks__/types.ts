import { vi } from "vitest";

export const App = vi.fn();
export const Component = vi.fn();
export const MarkdownRenderer = { render: vi.fn() };
export const Plugin = vi.fn();
