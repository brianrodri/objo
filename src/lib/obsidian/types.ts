export { Plugin } from "obsidian";
export type { EventRef } from "obsidian";
