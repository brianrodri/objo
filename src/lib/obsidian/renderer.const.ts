import { Duration } from "luxon";

export const MARKDOWN_RENDER_DEBOUNCE_TIME = Duration.fromDurationLike({ milliseconds: 500 });
