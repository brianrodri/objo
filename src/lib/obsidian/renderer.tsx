import { Duration } from "luxon";
import { App, Component, MarkdownRenderer } from "obsidian";
import { createElement, FunctionalComponent, JSX } from "preact";
import { useEffect, useRef } from "preact/hooks";
import { useDebounceCallback } from "usehooks-ts";

import { MARKDOWN_RENDER_DEBOUNCE_TIME } from "./renderer.const";

/**
 * Asynchronously renders the given markdown source code as Obsidian (the app) would.
 * Debounces values so that, ideally, renders have enough time to finish before being discarded for newer values.
 * @param props - configures how the markdown is rendered.
 * @returns component for asynchronously rendering the given markdown source code as Obsidian (the app) would.
 */
export const ObsidianMarkdown: FunctionalComponent<ObsidianMarkdownProps> = (props) => {
    const { app, component, markdown, sourcePath, tagName = "span", delay = MARKDOWN_RENDER_DEBOUNCE_TIME } = props;
    const elRef = useRef<HTMLElement>();
    const render = useDebounceCallback(MarkdownRenderer.render, delay.toMillis());

    useEffect(() => {
        const { current: el } = elRef;
        if (el) {
            render(app, markdown, el, sourcePath, component);
            return render.cancel;
        }
    }, [render, app, markdown, sourcePath, component]);

    return createElement(tagName, { ref: elRef });
};

/** Configures how Obsidian (the app) will render its markdown. */
export interface ObsidianMarkdownProps {
    /** A reference to the Obsidian app object. */
    app: App;
    /** A parent component to manage the lifecycle of the rendered child components. */
    component: Component;
    /** The Markdown source code. */
    markdown: string;
    /** The normalized path of this Markdown file, used to resolve relative internal links. */
    sourcePath: string;
    /** The tag used to contain the rendered Markdown source code. */
    tagName?: keyof JSX.IntrinsicElements;
    /** Custom debounce time for renders. */
    delay?: Duration<true>;
}
