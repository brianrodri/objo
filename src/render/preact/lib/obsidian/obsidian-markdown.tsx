import { Duration, DurationLike } from "luxon";
import { createElement, FunctionalComponent, JSX } from "preact";
import { useEffect, useMemo, useRef } from "preact/hooks";
import { useDebounceCallback } from "usehooks-ts";

import { App, Component, MarkdownRenderer } from "@/lib/obsidian/types";

import { MARKDOWN_RENDER_DEBOUNCE_TIME } from "./obsidian-markdown.const";

/**
 * @param props - props for the component.
 * @returns component that asynchronously renders the given markdown source code just as Obsidian (the app) would.
 */
export const ObsidianMarkdown: FunctionalComponent<ObsidianMarkdownProps> = (props) => {
    const { app, component, markdown, sourcePath, tagName = "span", delay = MARKDOWN_RENDER_DEBOUNCE_TIME } = props;
    const delayMs = useMemo(() => Duration.fromDurationLike(delay).toMillis(), [delay]);
    const renderObsidianMarkdown = useDebounceCallback(MarkdownRenderer["render"], delayMs);
    const elRef = useRef<HTMLElement>();

    useEffect(() => {
        const el = elRef.current;
        if (el) {
            const runAsync = async () => await renderObsidianMarkdown(app, markdown, el, sourcePath, component);
            runAsync().catch(console.error);
            return renderObsidianMarkdown.cancel;
        }
    }, [renderObsidianMarkdown, app, markdown, sourcePath, component]);

    return createElement(tagName, { ref: elRef });
};

/** Configures how Obsidian (the app) will render its markdown. */
export interface ObsidianMarkdownProps {
    /** A reference to the Obsidian app object. */
    app: App;
    /** A parent component to manage the lifecycle of the rendered child components. */
    component: Component;
    /** The Markdown source code. */
    markdown: string;
    /** The normalized path of this Markdown file, used to resolve relative internal links. */
    sourcePath: string;
    /** The tag used to contain the rendered Markdown source code. */
    tagName?: keyof JSX.IntrinsicElements;
    /** Custom debounce time for renders. */
    delay?: DurationLike;
}
