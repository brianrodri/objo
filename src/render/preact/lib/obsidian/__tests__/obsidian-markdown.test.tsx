import { render } from "@testing-library/preact";
import { DurationLike } from "luxon";
import { OmitByValue } from "utility-types";
import { afterAll, afterEach, beforeAll, describe, expect, it, vi } from "vitest";

import { App, Component, MarkdownRenderer } from "@/lib/obsidian/types";

import { ObsidianMarkdown, ObsidianMarkdownProps } from "../obsidian-markdown";

vi.mock("@/lib/obsidian/types");
afterEach(() => vi.restoreAllMocks());

describe(`${ObsidianMarkdown.name}`, () => {
    const requiredProps: OmitByValue<ObsidianMarkdownProps, undefined> = {
        app: new App(),
        component: new Component(),
        sourcePath: "/vault/diary.md",
        markdown: "Foo",
    };

    beforeAll(() => vi.useFakeTimers());
    afterAll(() => vi.useRealTimers());

    it("should render eventually", async () => {
        render(<ObsidianMarkdown {...requiredProps} />);
        expect(MarkdownRenderer["render"]).not.toHaveBeenCalled();
        await vi.runAllTimersAsync();
        expect(MarkdownRenderer["render"]).toHaveBeenCalledTimes(1);
    });

    it("should throw an error when 'props.delay' is invalid", () => {
        const invalid = "one hour" as DurationLike;
        expect(() => render(<ObsidianMarkdown {...requiredProps} delay={invalid} />)).toThrowErrorMatchingSnapshot();
    });

    it("should rerender when 'props.markdown' stops changing for a while", async () => {
        const { rerender } = render(<ObsidianMarkdown {...requiredProps} delay={500} markdown="Apple" />);
        await vi.runAllTimersAsync();
        rerender(<ObsidianMarkdown {...requiredProps} delay={500} markdown="Banana" />);
        await vi.advanceTimersByTimeAsync(300);
        rerender(<ObsidianMarkdown {...requiredProps} delay={500} markdown="Banana" />);
        await vi.advanceTimersByTimeAsync(300);

        expect(MarkdownRenderer["render"]).toHaveBeenCalledTimes(2);
        expect(MarkdownRenderer["render"]).toHaveBeenCalledWith(
            requiredProps.app,
            "Apple",
            expect.any(HTMLSpanElement),
            requiredProps.sourcePath,
            requiredProps.component,
        );
        expect(MarkdownRenderer["render"]).toHaveBeenCalledWith(
            requiredProps.app,
            "Banana",
            expect.any(HTMLSpanElement),
            requiredProps.sourcePath,
            requiredProps.component,
        );
    });

    it("should not rerender when 'props.markdown' keeps changing", async () => {
        const { rerender } = render(<ObsidianMarkdown {...requiredProps} delay={500} markdown="Apple" />);
        await vi.runAllTimersAsync();
        rerender(<ObsidianMarkdown {...requiredProps} delay={500} markdown="Banana" />);
        await vi.advanceTimersByTimeAsync(300);
        rerender(<ObsidianMarkdown {...requiredProps} delay={500} markdown="Cherry" />);
        await vi.advanceTimersByTimeAsync(300);

        expect(MarkdownRenderer["render"]).toHaveBeenCalledTimes(1);
        expect(MarkdownRenderer["render"]).toHaveBeenCalledWith(
            requiredProps.app,
            "Apple",
            expect.any(HTMLSpanElement),
            requiredProps.sourcePath,
            requiredProps.component,
        );
        expect(MarkdownRenderer["render"]).not.toHaveBeenCalledWith(
            requiredProps.app,
            "Banana",
            expect.any(HTMLSpanElement),
            requiredProps.sourcePath,
            requiredProps.component,
        );
        expect(MarkdownRenderer["render"]).not.toHaveBeenCalledWith(
            requiredProps.app,
            "Cherry",
            expect.any(HTMLSpanElement),
            requiredProps.sourcePath,
            requiredProps.component,
        );
    });
});
