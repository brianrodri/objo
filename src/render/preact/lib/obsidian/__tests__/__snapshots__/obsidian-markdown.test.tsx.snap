// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ObsidianMarkdown > should throw an error when 'props.delay' is invalid 1`] = `[Error: Unknown duration argument one hour of type string]`;
